### ISSUE TITLE
Improve "BeEquivalentTo" output for simple result sets

### ISSUE THREAD

**Issue by mdentremont** [fluentassertions/fluentassertions#1455]
(on latest version 5.10.3)

When `BeEquivalentTo` fails, I see the following output:
```
Expected subject to be a collection with 5 item(s), but {"global in ancestor", "global in current"}"
"contains 3 item(s) less than"
"{"global in ancestor", "global in current", "for current", "global elsewhere", "for somebody"}.
```

This is fine with smaller sets, but when things get large, it gets hard to read quickly.

Before we switched to Fluent Assertions, we custom assertion helpers that would log additional information like:
```
Subject is missing: {"for current", "global elsewhere", "for somebody"}
Subject has unexpected: {"some other value"}
```

Is this something that could easily be added, or would it necessitate some hacky changes?

Thanks!

**Comment by dennisdoomen** [comment]
I think it would be a great improvement. For smaller sets, you may want to see all of it. But if the collection gets too big, a subset might be better.

**Comment by mdentremont** [comment]
@dennisdoomen I can likely spend a bit of time on this in the somewhat near future - do you have any thoughts/tips on how to best achieve this (I took a quick peek and things look pretty optimized so I don't want to destroy the performance 😛)?

**Comment by dennisdoomen** [comment]
This would just be about finding the differences and properly reporting them. I suspect the changes would be limited to EnumerableEquivalencyValidatorExtensions

**Comment by JackDjTom** [comment]
This would be a great feature.
We're working also with large sets and it's always difficult to see the differences in the output window.

**Comment by iliashkolyar** [comment]
> This would just be about finding the differences and properly reporting them. I suspect the changes would be limited to EnumerableEquivalencyValidatorExtensions

I've looked at it a bit and I think it is a little more complicated then that.
Currently it seems that the `AssertCollectionHasEnoughItems` and `AssertCollectionHasNotTooManyItems` exist to perform a 'naive' comparison before prior to a "full equivalency" check, as they only compare the count of the asserted collection and the expected one.
If the counts are not similar, then we just print out the count difference.

As far as I understand, If we would like to actually print out the elements that are "missing" and "should not belong" from/to the asserted collection, we'll need to always run the `AssertElementGraphEquivalency`, which kind of makes the `HasEnoughItems`/`HasNotTooManyItems` checks irrelevant.
After all, elements could be "missing" from the asserted collection no matter if it has less/more or exactly the number of elements the expected collection.

So again, if I understand correctly, supporting this request means always running `AssertElementGraphEquivalency` and removing the current checks that reside in `EnumerableEquivalencyValidatorExtensions`.

What do you think?

**Comment by dennisdoomen** [comment]
You are right in your analysis. But I think we can simplify this by accepting that when the counts don't match, we just do a shallow comparison of the items in the collection and report the missing and unexpected items using `Equals`. 

BTW, your analysis triggered me to look at the implementation again. It turns out that the "naïve" check may not work as we intended it to work anymore. It was supposed to do a shallow comparison, but a while back we aligned all implementations of `BeEquivalentTo` so they would always use the object graph comparison.

**Comment by iliashkolyar** [comment]
Trying the shallow comparison leads to misleading results when working with non-primitive objects.
For example, if the new logic runs on the [the following test](https://github.com/fluentassertions/fluentassertions/blob/master/Tests/FluentAssertions.Specs/Equivalency/CollectionEquivalencySpecs.cs#L1995), we will receive a message that indicates that `Expected subject to be a collection with 1 item(s), but it contains 1 item(s) more`, and the "unexpected items" collection will hold both items in the `subject`.
This can be confusing for the caller since both the `subject` and the `expectation` hold a similar object:
```
new Customer
{
    Name = "John",
    Age = 27,
    Id = 1
}
```
But the shallow `Equals` comparison did not differentiate between them.

Do you think that this logic should be executed only on primitive collection types?
If so, was it already done somewhere else in the project (consistency-wise)?

**Comment by dennisdoomen** [comment]
You're right. This will be anything but obvious. The _proper_ solution is to also use the deep comparison that `EnumerableEquivalencyValidator` does to figure out which items are missing and with are extraneous. This would also take into account any configuration settings, especially when excluding/including specific members, collection ordering and rules that change the way complex and primitive types are compared. 

I think it's doable, but not trivial.

**Comment by iliashkolyar** [comment]
Thanks for the explanation.
I don't think i'll have the time to get into this with my current familiarity with the project.

Any chance we want a "quick win" by running shallow comparison only on primitive collection types?

**Comment by JackDjTom** [comment]
Can the [Debug.AssertInterpolatedStringHandler](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.debug.assertinterpolatedstringhandler?view=net-6.0) from .Net 6.0 be used for this scenario?

As far as I understand, you can create strings with this Handler, which will only be interpreted (allocate memory) when a certain condition is true. 
So maybe you could use that so the errormessage of BeEquivalentTo would only be created if the condition fails.

Here's an explanation video of this feature: [https://www.youtube.com/watch?v=24D7u3y3mhI](https://www.youtube.com/watch?v=24D7u3y3mhI)

**Comment by dennisdoomen** [comment]
We're still targeting .NET Framework 4.7, so no, not immediatelly.
